File: api/app/core/email/channel.py
"""The EmailChannel: compose + dispatch domain emails through a transport."""

from __future__ import annotations

from typing import Any

from app.core.cfo import DailyBrief
from app.core.email.renderer import render, render_daily_brief
from app.core.email.transport import Transport


class EmailChannel:
    def __init__(self, transport: Transport, *, sender: str = "cfo@maisha-mahsa.local") -> None:
        self._transport = transport
        self._sender = sender

    async def _send(self, *, to: str, subject: str, html: str) -> str:
        await self._transport.send(to=to, subject=subject, html=html, sender=self._sender)
        return html

    async def send_daily_brief(
        self, *, to: str, brief: DailyBrief, company_name: str = "Maisha-Mahsa"
    ) -> str:
        html = render_daily_brief(brief, company_name=company_name)
        flagged = len(brief.needs_attention)
        subject = f"{company_name} — Daily CFO Brief ({brief.as_of})" + (
            f" · {flagged} need(s) attention" if flagged else " · all green"
        )
        return await self._send(to=to, subject=subject, html=html)

    async def send_compliance_alert(self, *, to: str, ctx: dict[str, Any]) -> str:
        html = render("compliance_alert.html", ctx=ctx)
        subject = f"Compliance Alert · {ctx['total']} filing(s) need attention"
        return await self._send(to=to, subject=subject, html=html)

    async def send_payroll_approval(self, *, to: str, ctx: dict[str, Any]) -> str:
        html = render("payroll_approval.html", ctx=ctx)
        subject = f"Payroll Approval · {ctx['month_year']} · {ctx['validation_status'].upper()}"
        return await self._send(to=to, subject=subject, html=html)

    async def send_investor_update(
        self, *, to: str, ctx: dict[str, Any], company_name: str = "Maisha-Mahsa"
    ) -> str:
        html = render("investor_update.html", ctx=ctx, company_name=company_name)
        subject = f"{company_name} — Investor Update · {ctx['period']}"
        return await self._send(to=to, subject=subject, html=html)

File: api-nest/src/email/channel.ts
/** The EmailChannel: compose + dispatch domain emails through a transport. Port of channel.py. */
import { BriefPayload } from '../cfo/cfo';
import {
  renderComplianceAlert,
  renderDailyBrief,
  renderDunning,
  renderInvestorUpdate,
  renderPayrollApproval,
} from './render';
import { Transport } from './transport';

export class EmailChannel {
  constructor(
    private readonly transport: Transport,
    private readonly sender = 'cfo@maisha-mahsa.local',
  ) {}

  private async send(to: string, subject: string, html: string): Promise<string> {
    await this.transport.send({ to, subject, html, sender: this.sender });
    return html;
  }

  async sendDailyBrief(to: string, brief: BriefPayload, companyName = 'Maisha-Mahsa'): Promise<string> {
    const html = renderDailyBrief(brief, companyName);
    const flagged = brief.needs_attention.length;
    const subject =
      `${companyName} — Daily CFO Brief (${brief.as_of})` +
      (flagged ? ` · ${flagged} need(s) attention` : ' · all green');
    return this.send(to, subject, html);
  }

  async sendComplianceAlert(to: string, ctx: Record<string, any>): Promise<string> {
    const html = renderComplianceAlert(ctx);
    return this.send(to, `Compliance Alert · ${ctx.total} filing(s) need attention`, html);
  }

  async sendDunning(to: string, ctx: Record<string, any>, companyName = 'Maisha-Mahsa'): Promise<string> {
    const html = renderDunning(ctx, companyName);
    return this.send(to, `Payment reminder · Invoice ${ctx.invoice_number} (${ctx.stage})`, html);
  }

  async sendPayrollApproval(to: string, ctx: Record<string, any>): Promise<string> {
    const html = renderPayrollApproval(ctx);
    return this.send(to, `Payroll Approval · ${ctx.month_year} · ${String(ctx.validation_status).toUpperCase()}`, html);
  }

  async sendInvestorUpdate(to: string, ctx: Record<string, any>, companyName = 'Maisha-Mahsa'): Promise<string> {
    const html = renderInvestorUpdate(ctx, companyName);
    return this.send(to, `${companyName} — Investor Update · ${ctx.period}`, html);
  }
}
